library/pkparse.c

```diff
@@ -1082,7 +1082,10 @@ int mbedtls_pk_parse_key( mbedtls_pk_context *pk,
     if( ret == 0 )
     {
         if( ( pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_RSA ) ) == NULL )
+        {
+            mbedtls_pem_free( &pem );
```

Comment:
I know that test was already there, but can this code path really be reached? We're inside `#ifdef MBEDTLS_RSA_C`. I think I'd rather not check `pk_info` here and just call `pk_setup()` directly, as it checks `pk_info` anyway, but at least we avoid unreachable branches in the code.

Reply:
@mpg I agree, the code-path should be unreachable, and as `pk` is required to be initialized or freed at function-entry, it should be safe to call `mbedtls_pk_free` on it even if it hasn't been modified at all. So I think we can omit the check for `pk_info != NULL`.

Reply:
@mpg @andresag01 Could you also have a look at https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L1220 and https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L1233 please? The logic seems strange there: If for whatever reasons it happens that `mbedtls_pk_setup` fails, the branch will be taken and `0` is returned. I suggest to change this to
```
#if defined(MBEDTLS_RSA_C)

    pk_info = mbedtls_pk_info_from_type( MBEDTLS_PK_RSA );
    if( ( ret = mbedtls_pk_setup( pk, pk_info ) ) != 0 ||
        ( ret = pk_parse_key_pkcs1_der( mbedtls_pk_rsa( *pk ),
                                        key, keylen ) ) != 0 )
    {
        mbedtls_pk_free( pk );
    }
    else
    {
        return( 0 );
    }

#endif /* MBEDTLS_RSA_C */
```

Reply:
@hanno-arm: Well spotted, here is my analysis:
* https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L1220: I think  its broken, if `mbedtls_pk_setup()` fails we should probably just return an error. In any case, with the current code, I do not think we will ever succeed to parse those keys because setup() would have failed...
* https://github.com/ARMmbed/mbedtls/blob/development/library/pkparse.c#L1233: Same analysis as above.

Reply:
@mpg @andresag01 Added a fix.